dnsprovider/pkg/dnsprovider/providers/scaleway/dns.go

```diff
@@ -0,0 +1,571 @@
+	token := &oauth2.Token{
+		AccessToken: t.AccessToken,
+	}
+	return token, nil
+}
+
+func newClient() (*scw.Client, error) {
+	accessToken := os.Getenv("SCW_ACCESS_KEY")
+	if accessToken == "" {
+		return nil, errors.New("SCW_ACCESS_KEY is required")
+	}
+	tokenSource := &TokenSource{
+		AccessToken: accessToken,
+	}
+	oauthClient := oauth2.NewClient(context.TODO(), tokenSource)
+
+	// TODO: check if it's necessary to have both the oauth token & the scw env
+	scwClient, err := scw.NewClient(
+		scw.WithHTTPClient(oauthClient),
+		scw.WithUserAgent("kubernetes-kops/"+kopsv.Version),
+		scw.WithEnv(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return scwClient, nil
+}
+
+// Interface implements dnsprovider.Interface
+type Interface struct {
+	client       *scw.Client
+	parentDomain string
+}
+
+// NewProvider returns an implementation of dnsprovider.Interface
+func NewProvider(client *scw.Client, parentDomain string) dnsprovider.Interface {
+	return &Interface{client: client, parentDomain: parentDomain}
+}
+
+// Zones returns an implementation of dnsprovider.Zones
+func (d Interface) Zones() (dnsprovider.Zones, bool) {
+	return &zones{
+		client:       d.client,
+		parentDomain: d.parentDomain,
+	}, true
+}
+
+// zones is an implementation of dnsprovider.Zones
+type zones struct {
+	client       *scw.Client
+	parentDomain string
+}
+
+// List returns a list of all dns zones
+func (z *zones) List() ([]dnsprovider.Zone, error) {
+	domains, err := listDomains(z.client)
+	if err != nil {
+		return nil, err
+	}
+
+	var newZone *zone
+	var zones []dnsprovider.Zone
+	for _, domainSummary := range domains {
+		newZone = &zone{
+			name:         domainSummary.Domain,
+			parentDomain: z.parentDomain,
+			client:       z.client,
+		}
+		zones = append(zones, newZone)
+	}
+
+	return zones, nil
+}
+
+// Add adds a new DNS zone
+func (z *zones) Add(newZone dnsprovider.Zone) (dnsprovider.Zone, error) {
+	domainCreateRequest := &domain.CreateDNSZoneRequest{
+		Subdomain: newZone.Name(),
+		Domain:    z.parentDomain,
+	}
+
+	d, err := createDomain(z.client, domainCreateRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	return &zone{
+		name:         d.Subdomain,
+		parentDomain: d.Domain,
+		client:       z.client,
+	}, nil
+}
+
+// Remove deletes a zone
+func (z *zones) Remove(zone dnsprovider.Zone) error {
+	return deleteDomain(z.client, zone.Name()+"."+z.parentDomain)
+}
+
+// New returns a new implementation of dnsprovider.Zone
+func (z *zones) New(name string) (dnsprovider.Zone, error) {
+	return &zone{
+		name:         name,
+		parentDomain: z.parentDomain,
+		client:       z.client,
+	}, nil
+}
+
+// zone implements dnsprovider.Zone
+type zone struct {
+	name         string
+	client       *scw.Client
+	parentDomain string
+	//id           string
+}
+
+// Name returns the Name of a dns zone
+func (z *zone) Name() string {
+	return z.name
+}
+
+// ID returns the ID of a dns zone
+func (z *zone) ID() string {
+	//return z.id
```

Comment:
Depends on what the domain API is taking in argument. If you can fetch the data of a zone with just its name, let's stick with the name as it allows for easier debugging. But if it is not possible and the ID is required during a get operation, let's use the ZoneID as a identifier.